QueryBuilder/Common/Statements/WhereStatement.cs

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Statements
+{
+    using System;
+    using System.Diagnostics.CodeAnalysis;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.Serialization;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Clauses;
+    using Microsoft.DigitalWorkplace.DigitalTwins.QueryBuilder.Common.Helpers;
+
+    /// <summary>
+    /// An abstract class that provides the base layer of methods that are common between
+    /// WHERE statement classes that are applicable to either Twins or Relationships.
+    /// </summary>
+    /// <typeparam name="TWhereStatement">The type of WHERE statement supported.</typeparam>
+    public abstract class WhereBaseStatement<TWhereStatement> where TWhereStatement : WhereBaseStatement<TWhereStatement>
```

Comment:
Because we agreed there was some semantics added in the dynamic builder that we would like to eventually carry into the typed one, thus I felt like to achieved that, the typed one would likely need some typed version of these that reuses the current ones in some way to avoid too much duplication. If we feel that it would likely just be too different considering what the typed queries bring into the equation, then I can move them out of Common.